cpp/src/io/parquet/reader_impl.cu

```diff
@@ -278,38 +278,48 @@ struct metadata : public FileMetaData {
    *
    * @param row_group Index of the row group to select
    * @param max_rowgroup_count Max number of consecutive row groups if > 0
+   * @param row_group_indices Arbitrary rowgroup list[max_rowgroup_count] if non-null
    * @param row_start Starting row of the selection
    * @param row_count Total number of rows selected
    *
    * @return List of row group indexes and its starting row
    */
-  auto select_row_groups(int row_group, int max_rowgroup_count, int &row_start, int &row_count) {
-    std::vector<std::pair<int, int>> selection;
+  auto select_row_groups(size_type row_group, size_type max_rowgroup_count,
+                         const size_type *row_group_indices,
+                         size_type &row_start, size_type &row_count) {
+    std::vector<std::pair<size_type, size_t>> selection;
 
-    if (row_group != -1) {
-      CUDF_EXPECTS(row_group < get_num_row_groups(), "Non-existent row group");
-      for (int i = 0; i < row_group; ++i) {
-        row_start += row_groups[i].num_rows;
+    if (row_group_indices) {
+      row_count = 0;
+      for (size_type i = 0; i < max_rowgroup_count; i++) {
+        auto rowgroup_idx = row_group_indices[i];
+        CUDF_EXPECTS(rowgroup_idx >= 0 && rowgroup_idx < get_num_row_groups(), "Invalid rowgroup index");
+        selection.emplace_back(rowgroup_idx, row_count);
+        row_count += row_groups[rowgroup_idx].num_rows;
       }
+    }
+    else if (row_group != -1) {
+      CUDF_EXPECTS(row_group < get_num_row_groups(), "Non-existent row group");
       row_count = 0;
       do {
         selection.emplace_back(row_group, row_start + row_count);
         row_count += row_groups[row_group].num_rows;
       } while (--max_rowgroup_count > 0 && ++row_group < get_num_row_groups());
     } else {
       row_start = std::max(row_start, 0);
-      if (row_count == -1) {
-        row_count = get_total_rows();
+      if (row_count < 0) {
+        row_count = static_cast<size_type>(std::min<int64_t>(get_total_rows(), INT32_MAX));
       }
       CUDF_EXPECTS(row_count >= 0, "Invalid row count");
       CUDF_EXPECTS(row_start <= get_total_rows(), "Invalid row start");
 
-      for (int i = 0, count = 0; i < (int)row_groups.size(); ++i) {
+      for (size_t i = 0, count = 0; i < row_groups.size(); ++i) {
+        size_t chunk_start_row = count;
         count += row_groups[i].num_rows;
-        if (count > row_start || count == 0) {
-          selection.emplace_back(i, count - row_groups[i].num_rows);
+        if (count > static_cast<size_t>(row_start) || count == 0) {
```

Comment:
`count` and `row_start` are both `size_t`
```suggestion
        if (count > row_start || count == 0) {
```

Reply:
Again, no: count is size_t, but row_start is size_type